pkg/cmd/pulumi/packagecmd/project_sdk_link.go

```diff
@@ -846,80 +868,150 @@ func ProviderFromSource(pctx *plugin.Context, packageSource string) (plugin.Prov
+			return Provider{}, err
+		}
 
-			log := func(sev diag.Severity, msg string) {
-				pctx.Host.Log(sev, "", msg, 0)
-			}
+		log := func(sev diag.Severity, msg string) {
+			pctx.Host.Log(sev, "", msg, 0)
+		}
 
-			_, err = pkgWorkspace.InstallPlugin(pctx.Base(), descriptor.PluginSpec, log)
-			if err != nil {
-				return nil, err
-			}
+		_, err = pkgWorkspace.InstallPlugin(pctx.Base(), descriptor.PluginSpec, log)
+		if err != nil {
+			return Provider{}, err
+		}
+
+		p, err = pctx.Host.Provider(descriptor)
+		if err != nil {
+			return Provider{}, err
+		}
 
-			p, err := pctx.Host.Provider(descriptor)
+		return Provider{Provider: p}, nil
+	}
+
+	setupProvider := func(
+		descriptor workspace.PackageDescriptor, specOverride *workspace.PackageSpec,
+	) (Provider, *workspace.PackageSpec, error) {
+		p, err := installDescriptor(descriptor)
+		if err != nil {
+			return Provider{}, nil, err
+		}
+		if descriptor.Parameterization != nil {
+			_, err := p.Provider.Parameterize(pctx.Request(), plugin.ParameterizeRequest{
+				Parameters: &plugin.ParameterizeValue{
+					Name:    descriptor.Parameterization.Name,
+					Version: descriptor.Parameterization.Version,
+					Value:   descriptor.Parameterization.Value,
+				},
+			})
 			if err != nil {
-				return nil, err
+				return Provider{}, nil, fmt.Errorf("failed to parameterize %s: %w", p.Provider.Pkg().Name(), err)
 			}
+		}
+		return p, specOverride, nil
+	}
 
-			return p, nil
+	// For all plugins except for file paths, we try to load it from the provider host.
+	//
+	// Note that if a local folder has a name that could match an downloadable plugin, we
+	// prefer the downloadable plugin.  The user can disambiguate by prepending './' to the
+	// name.
+	if !plugin.IsLocalPluginPath(pctx.Base(), packageSource) {
+		if !env.DisableRegistryResolve.Value() && env.Experimental.Value() {
+			meta, err := registry.ResolvePackageFromName(pctx.Base(), reg, pluginSpec.Name, pluginSpec.Version)
+			if err == nil {
+				spec := workspace.PluginSpec{
+					Name:              meta.Name,
+					Kind:              apitype.ResourcePlugin,
+					Version:           &meta.Version,
+					PluginDownloadURL: meta.PluginDownloadURL,
+				}
+				var params *workspace.Parameterization
+				if meta.Parameterization != nil {
+					spec.Name = meta.Parameterization.BaseProvider.Name
+					spec.Version = &meta.Parameterization.BaseProvider.Version
+					params = &workspace.Parameterization{
+						Name:    meta.Name,
+						Version: meta.Version,
+						Value:   meta.Parameterization.Parameter,
+					}
+				}
+				return setupProvider(workspace.NewPackageDescriptor(spec, params), &workspace.PackageSpec{
+					Source:  meta.Source + "/" + meta.Publisher + "/" + meta.Name,
+					Version: meta.Version.String(),
+				})
+			} else if errors.Is(err, registry.ErrNotFound) &&
+				registry.PulumiPublishedBeforeRegistry(pluginSpec.Name) {
+				// Let's try installing it without the registry.
+			} else {
```

Comment:
Hrm wouldn't we get here as well e.g. if no registry is available?  What does `GetSuggestedPackages` return in that case?

Reply:
If no registry were available, then `GetSuggestedPackages` would return `nil`. `GetSuggestedPackages` just looks at the passed in error, it doesn't make any API calls.